forem/forem
# Part 2: Shows Relevant Posts in Homepage Sidebar

## What type of PR is this? (check all applicable)


## Description
The "Definition of Done" for this RFC is:

> This feature is deployed to all Forems, and we can assert that every size Forem, regardless of Forem size, can render articles on the sidebar. We should be able to see that every Forem (even if those that have few posts or old posts) can always display posts for a tag, regardless of an article's `published_at` date or `score`.

This PR is part two of a two-part solution to ensure that posts show beneath the tags in the homepage's righthand sidebar, regardless of the Forem's size and the article's `published_at` date and `score`. The second part of this two-part solution focuses specifically on all active tags (including the `#discuss` tag) other than the `#help` tag, in the homepage's righthand sidebar. To accomplish this, I've removed the `active_threads` class method from the `Article` model and replaced it with `Articles::ActiveThreadsQuery` to follow the query object pattern. This is responsible for determining which posts to show under the active tags found in the homepage's righthand sidebar. In addition to this, I've replaced all `.active_threads` calls with `Articles::ActiveThreadsQuery.call(options: {...})` in the necessary views. Finally, I've added a test for this work to `spec/queries/articles/active_threads_query_spec.rb` and removed the tests from `spec/models/article_spec.rb`.

## Related Tickets & Documents
Relates to and closes [RFC 20](https://github.com/forem/rfcs/pull/20)

## QA Instructions, Screenshots, Recordings
**To QA this PR, please first ensure that you have admin-status and then navigate to `/admin/config` and scroll to the `/tags` section**:
<img width="1037" alt="107829988-22920580-6d48-11eb-89c9-5a0c50cb3f3f" src="https://user-images.githubusercontent.com/32834804/110387534-f7d15f00-801e-11eb-8a7a-e4edd87025a6.png">

- In order to have tags display in the sidebar and to test this PR, you need to first set some tags, like "discuss" and "watercooler" (⚠️ in order to test this, you will need to set the "discuss" tag!):
<img width="974" alt="Screen Shot 2021-03-08 at 3 02 33 PM" src="https://user-images.githubusercontent.com/32834804/110387750-526abb00-801f-11eb-8f61-05842a19a928.png">

- After setting the tags, navigate to the homepage and observe that there are no articles beneath the "watercooler" tag and that there are some posts beneath the "discuss" tag (these posts meet the criteria within `Articles::ActiveThreadsQuery`):
<img width="1128" alt="Screen Shot 2021-03-08 at 3 12 51 PM" src="https://user-images.githubusercontent.com/32834804/110388734-c5286600-8020-11eb-9e16-c1725b4a1e0f.png">

- In order to see articles (other than the ones created when you `seed` your database!) displayed beneath the "discuss" and "watercooler" tags, create a post that uses both tags:
<img width="892" alt="Screen Shot 2021-03-08 at 3 22 54 PM" src="https://user-images.githubusercontent.com/32834804/110389723-2997f500-8022-11eb-8897-612bfa7e11df.png">
<img width="910" alt="Screen Shot 2021-03-08 at 3 16 18 PM" src="https://user-images.githubusercontent.com/32834804/110389084-3d8f2700-8021-11eb-957b-7c90c42577c8.png">

- After creating articles with the "discuss" and "watercooler" tags and ensuring that they are lower-quality and do not fall within the constraints of `Articles::ActiveThreadsQuery`, navigate back to the homepage. Observe that your articles tagged with "discuss" and "watercooler" display in the sidebar:
![Screen Shot 2021-03-18 at 12 00 37 PM](https://user-images.githubusercontent.com/32834804/111674380-8ee9a400-87e1-11eb-90e1-1d232674b43f.png)
![Screen Shot 2021-03-18 at 12 01 24 PM](https://user-images.githubusercontent.com/32834804/111674485-ac1e7280-87e1-11eb-8c22-972333dd8031.png)

- Try to break things! ⚒️

### UI accessibility concerns?
None!

## Added tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
✅ I updated the Admin Guide in the first part of this two-part work!
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post-deployment tasks we need to perform?
This will need to be deployed to all Forems.

## [optional] What gif best describes this PR or how it makes you feel?

![New Post Gif](https://media.giphy.com/media/f61UuIFFm3MLMkds5b/giphy.gif)
